MBS-7678: Combine display of multiple relationships to the same entity

mwiencek opened the pull request:
https://tickets.metabrainz.org/browse/MBS-7678

mwiencek commented on root/static/scripts/common/i18n/semicolonOnlyList.js:
```diff
@@ -0,0 +1,44 @@
+/*
+ * @flow
+ * Copyright (C) 2015-2016 MetaBrainz Foundation
```
I kept the copyright dates from `commaOnlyList` since it's basically 99% derived from that, not really anything new.

Re changing `commaOnlyList` to take an optional separator, it's possible, but it'd have to be a hardcoded list of separators (since it's fixed/set in the translated string), so we'd need a switch statement to pick the correct string and pass the correct interpolation argument names (semicolon_only_list_item vs. comma_only_list_item). TBH it also feels weird to allow that since "comma" is explicitly in the name (maybe we should just rename the function then, but).

mwiencek commented on root/static/scripts/common/i18n/expand2.js:
```diff
@@ -290,7 +297,9 @@ export default function expand<+T, V>(
-  if (!(args instanceof VarArgs)) {
+  if (!args ||
+      typeof args.has !== 'function' ||
+      typeof args.get !== 'function') {
```
I added an `implementsVarArgsInterface` function to hopefully clarify this part.

mwiencek commented on root/static/scripts/edit/utility/linkPhrase.js:
```diff
@@ -71,6 +71,12 @@ class PhraseVarArgs<T> implements VarArgsInterface<T | string> {
+   * phrase. Later used for calculating "extra" attributes (which
+   * didn't appear in the link phrase, so that we can display them
+   * separately).
+   */
```
Agreed! Renamed it to `usedPhraseAttributes`.

mwiencek commented on root/static/scripts/edit/utility/linkPhrase.js:
```diff
@@ -6,11 +6,10 @@
  * http://www.gnu.org/licenses/gpl-2.0.txt
  */
 
+import identity from 'lodash/identity';
```
Nothing, I must've forgotten to remove it during development. Gone now

mwiencek commented on root/utility/groupRelationships.js:
```diff
@@ -111,17 +248,160 @@ export default function groupRelationships(
+       * to the same target but with different link orders, we use the link
+       * order for grouping just so that the state of the data is visible to
+       * the user, allowing them to fix it if needed.
+       */
```
Good to know, thanks. In that case we'd definitely want to include the link order, so I just changed the comment entirely.

mwiencek commented:
> https://musicbrainz.org/release/2e327023-ec02-3650-82fa-5fa79ddaa08a has "acoustic guitar", "electric guitar [Lead Guitar]" and "electric guitar [Rhythm Guitar]" credited to the same two people - are we also considering grouping those?

Well firstly, I hadn't made the inline tracklist credits use this at all, since that wasn't converted to React yet. But that's done now in the latest commit.

About merging the "phrase groups" together, I hadn't considered that since it wasn't mentioned in the ticket, but I think it's a cool idea, so I've also amended the main commit to do that. It only merges two phrase groups together if they share the same link type or parent link type, and have the exact same lists of target entities (including associated dates/extra attributes).

> Also, while "Kirk Hammett (background vocals, guitar) (1983-04-01 –)" isn't too confusing, I feel "Kirk Hammett (1983-04-01 –: background vocals, guitar)" is - the "–:" is pretty strange. Maybe we should specify "present" there, or something?

I changed it to use root/utility/relationshipDateText.js, but now it might look too verbose. :) Let me know what you think. (I'm not sure we want to have three different ways of rendering relationship dates though, unless you think we should change the "compact" date period formatting across the entire site to be like that.)

reosarevok commented:
> I changed it to use root/utility/relationshipDateText.js, but now it might look too verbose. :)

I actually like it, but we should ask users. Maybe put a screenshot or two up here and let's point people on IRC to it?

> About merging the "phrase groups" together, I hadn't considered that since it wasn't mentioned in the ticket, but I think it's a cool idea, so I've also amended the main commit to do that. It only merges two phrase groups together if they share the same link type or parent link type, and have the exact same lists of target entities (including associated dates/extra attributes).

I see release/2e327023-ec02-3650-82fa-5fa79ddaa08a now has "engineer and co-producer", but then the same people are separately linked as "recording engineer". I guess that's because "recording engineer" is under "engineer", not with it? It does feel odd to have "mixer and mastering: Andy Sneap" separate from "recording engineer: Andy Sneap" though. I guess that is because there's two recording engineers?

mwiencek commented:
Hopefully resolved the "recording engineer" issue, and tried to simplify the implementation a bit (though it's still complicated, JavaScript just isn't a nice language to do this sort of data manipulation in).

> I actually like it, but we should ask users. Maybe put a screenshot or two up here and let's point people on IRC to it?

Here's one for the NIN members list:

<img width="644" alt="mbs-7678-verbose-dates" src="https://user-images.githubusercontent.com/1056556/69994955-bc3f0f80-1514-11ea-956e-ed55ab78555d.png">

> I see release/2e327023-ec02-3650-82fa-5fa79ddaa08a now has "engineer and co-producer", but then the same people are separately linked as "recording engineer". I guess that's because "recording engineer" is under "engineer", not with it?

Yeah, I had to change the logic to group things if they shared the same link type root, otherwise couldn't see a way to get this to work. So now it looks like this:

<img width="577" alt="mbs-7678-phrase-group-merging" src="https://user-images.githubusercontent.com/1056556/69995054-f5777f80-1514-11ea-9ecb-6c846f56657f.png">

> It does feel odd to have "mixer and mastering: Andy Sneap" separate from "recording engineer: Andy Sneap" though. I guess that is because there's two recording engineers?

Right. The alternative would be

```
mixer, mastering, engineer, co-producer and recording engineer: Andy Sneap
engineer, co-producer and recording engineer: Jeff Balding
```
but this looks worse to me. There are of course other ways to divide those up, but it might be hard to come up with a consistent rule for doing that that looks good in all cases.

mwiencek commented:
I rebased this, fixed the conflicts, and cleaned up some unused code I noticed. Also removed the commit adding the `VarArgsInterface` stuff, since I found a nicer way to work with what we have. So that simplifies things slightly too.

mwiencek commented:
Rebased again and tested that it still works.

reosarevok commented on root/static/scripts/common/linkedEntities.js:
```diff
@@ -15,22 +15,23 @@
 type LinkedEntities = {
   artist_type: {
-    +[number]: ArtistTypeT,
+    [number]: ArtistTypeT,
```
Not newly added, ofc, but these probably should be id: number now, even if the eslint rule doesn't detect them.

mwiencek replied:
Pushed a commit to add those.

reosarevok commented on root/static/scripts/common/i18n/semicolonOnlyList.js:
```diff
@@ -0,0 +1,44 @@
+
+export default semicolonOnlyList;
+
+export {semicolonOnlyListText};
```
Isn't `export const x` clearer? We use that much more often than `export {x}` at the end (but we do use it this way a few times, so i guess it's not a big deal).

mwiencek replied:
Both seem fine to me but I simplified it by moving the `export default` to the `semicolonOnlyList` declaration, and then just doing `export function semicolonOnlyListText(...)` for parity.
